android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/android/service/DefaultGreengrassComponentService.java

```diff
@@ -0,0 +1,375 @@
+                    } else {
+                        throw new RuntimeException("Startup attempts limit reached");
+                    }
+                } else if (ACTION_START_COMPONENT.equals(intent.getAction())) {
+                    // Do normal startup if everything is fine
+                    startAttemptsCounter++;
+                    super.onStartCommand(intent, flags, startId);
+                    return START_STICKY;
+                } else {
+                    // Unknown action - ignore the intent
+                    throw new InvalidArgumentsError("Unsupported action in start intent: "
+                            + intent.getAction());
+                }
+            } catch (Throwable e) {
+                if (logger != null) {
+                    logger.atError().setCause(e).log("Fatal error at startup");
+                }
+
+                // Abort startup by reporting not-sticky start
+                stopSelf();
+                return START_NOT_STICKY;
+            }
+        }
+    }
+
+    private class NucleusWorkerThread extends ComponentWorkerThread {
+
+        @Override
+        public void run() {
+            // This is the protection from malformed counter
+            synchronized (startAttemptsCounter) {
+                if (startAttemptsCounter > NUCLEUS_START_ATTEMPTS_LIMIT) {
+                    logger.atError()
+                            .log("Startup attempts counter is over the limit. "
+                                    + "Probably, startup intent is malformed. Startup aborted");
+                    workerExitCode = EXIT_CODE_FAILED;
+                    return;
+                }
+            }
+
+            try {
+                // Enter critical section to avoid sudden interruption of Nucleus startup process
+                enterCriticalSection();
+                logger.atDebug().log("Entered critical section for Nucleus startup");
+
+                AndroidPlatform platform = (AndroidPlatform) Platform.getInstance();
+                platform.setAndroidAPIs(DefaultGreengrassComponentService.this, packageManager, componentManager);
+
+                ProvisionManager provisionManager = BaseProvisionManager.getInstance(getFilesDir());
+                final String[] nucleusArguments = provisionManager.prepareArguments();
+                kernel = GreengrassSetup.main(nucleusArguments);
+
+                // Clear system properties
+                provisionManager.clearSystemProperties();
+
+                if (!provisionManager.isProvisioned()) {
+                    provisionManager.writeConfig(kernel);
+                }
+
+                // Startup is done, now we can leave critical section and allow to terminate Nucleus
+                leaveCriticalSection();
+                logger.atDebug().log("Nucleus startup is done, left critical section");
+
+                // waiting for Thread.interrupt() call
+                join();
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)